### PR TITLE
avoid using LocalFree on FormatMessageA buffer

### PR DESCRIPTION
**Pull request by orilevari**
**Description**: Pass in preallocated buffer to FormatMessageA to avoid using LocalFree.

**Motivation and Context**
LocalFree is introduced in Windows 10. Since 1.2 release, LoopDir which ends up calling LocalFree is being compiled into production code, not just test code so this breaks downlevel support contracts for WinML and OnnxRuntime.

### CODE REVIEW

File: onnxruntime/core/platform/path_lib.h [hunk 1]
```diff
@@ -178,11 +178,10 @@ inline OrtFileType DTToFileType(DWORD dwFileAttributes) {
   return OrtFileType::TYPE_REG;
 }
 inline std::string FormatErrorCode(DWORD dw) {
-  char* lpMsgBuf;
-  FormatMessageA(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS, NULL, dw,
-                 MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), (LPSTR)&lpMsgBuf, 0, NULL);
+  char lpMsgBuf[1024 * sizeof(TCHAR)];
```
Comment (snnn) [inline onnxruntime/core/platform/path_lib.h:181]: How to ensure it's enough?

Comment (orilevari) [reply onnxruntime/core/platform/path_lib.h:181]: It doesn't seem like there is a way. Looking at Windows system error codes, seems like the vast majority will fit in this amount of bytes though. And if its not enough, the docs make it seem like the returned buffer will still be null terminated.

**Comment by tiagoshibata** [comment]
It's good enough for me 👍 I'll wait for @snnn 's review though

File: onnxruntime/core/platform/path_lib.h (hunk 1 above)
Comment (martinb35) [inline onnxruntime/core/platform/path_lib.h:181]: in general its better to use a constant instead of hard-coding '1024' twice.

Comment (martinb35) [reply onnxruntime/core/platform/path_lib.h:181]: from the doc (https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-formatmessage): 
If the FORMAT_MESSAGE_ALLOCATE_BUFFER flag is not set, this parameter specifies the size of the output buffer, in TCHARs. If FORMAT_MESSAGE_ALLOCATE_BUFFER is set, this parameter specifies the minimum number of TCHARs to allocate for an output buffer.

The output buffer cannot be larger than 64K bytes.

Comment (orilevari) [reply onnxruntime/core/platform/path_lib.h:181]: Should I allocate a buffer of 64K? Seems like a lot of memory to use for messages that look like this: https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes--0-499-

Comment (orilevari) [reply onnxruntime/core/platform/path_lib.h]: Actually, I just tested to see how the api would behave for error messages that don't fit into the buffer and it returns a corrupted string. So I'll go ahead and allocate a 64K buffer on the heap.

File: onnxruntime/core/platform/path_lib.h [hunk 2]
```diff
@@ -178,11 +178,12 @@ inline OrtFileType DTToFileType(DWORD dwFileAttributes) {
...
 }
 inline std::string FormatErrorCode(DWORD dw) {
-  char* lpMsgBuf;
-  FormatMessageA(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS, NULL, dw,
-                 MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), (LPSTR)&lpMsgBuf, 0, NULL);
+  static const DWORD bufferLength = 64 * 1024;
+  char * lpMsgBuf = new char[bufferLength];
+  FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS, NULL, dw,
+                 MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), (LPSTR)lpMsgBuf, bufferLength / sizeof(TCHAR), NULL);
   std::string s(lpMsgBuf);
```
Comment (snnn) [inline onnxruntime/core/platform/path_lib.h:185]: Nit: you may use the buffer from s to reduce one heap allocation.
```
std::string s(64);
FormatMessageA(...,(LPSTR)s.data(),...)
```

Comment (orilevari) [reply onnxruntime/core/platform/path_lib.h]: Went ahead and changed this, can you reapprove? thanks!

**Comment by snnn** [comment]
Thank you!

**Comment by snnn** [comment]
Sorry, I forgot this. I introduced it back again in PR #9975 ... I will submit a fix in the next week.

**Pull request merged**